Possible race condition in htex scale down

**Describe the bug**
I've seen the below test failure as a one-off in `--config local` tests in the CI on several occasions.

My gut feeling is there's a race condition here - either in the test, or in parsl itself. I'm noting this for later investigation.

The scale down behaviour here is relatively new. Tagging @yadudoc who implemented it.

```
parsl/tests/test_scaling/test_scale_down.py F
=================================== FAILURES ===================================
________________________________ test_scale_out ________________________________
    @pytest.mark.local
    def test_scale_out():
        dfk = parsl.dfk()
    
        # Since we have init_blocks = 0, at this point we should have 0 managers
        print("Executor : ", dfk.executors['htex_local'])
        print("Before")
        print("Managers   : ", dfk.executors['htex_local'].connected_managers)
        print("Outstanding: \n", dfk.executors['htex_local'].outstanding)
        assert len(dfk.executors['htex_local'].connected_managers) == 0, "Expected 0 managers at start"
    
        fus = [sleeper(i) for i in [3, 3, 25, 25, 50]]
    
        for i in range(2):
            fus[i].result()
    
        # At this point, since we have 1 task already processed we should have atleast 1 manager
        print("Between")
        print("Managers   : ", dfk.executors['htex_local'].connected_managers)
        print("Outstanding: \n", dfk.executors['htex_local'].outstanding)
>       assert len(dfk.executors['htex_local'].connected_managers) == 5, "Expected 5 managers once tasks are running"
E       AssertionError: Expected 5 managers once tasks are running
E       assert 3 == 5
E        +  where 3 = len([{'active': True, 'block_id': '1', 'idle_duration': 0, 'manager': '884d27969893', ...}, {'active': True, 'block_id': '...486f', ...}, {'active': True, 'block_id': '2', 'idle_duration': 0.0022573471069335938, 'manager': 'a19000850efd', ...}])
E        +    where [{'active': True, 'block_id': '1', 'idle_duration': 0, 'manager': '884d27969893', ...}, {'active': True, 'block_id': '...486f', ...}, {'active': True, 'block_id': '2', 'idle_duration': 0.0022573471069335938, 'manager': 'a19000850efd', ...}] = HighThroughputExecutor(\n    address=None, \n    address_probe_timeout=None, \n    cores_per_worker=1.0, \n    cpu_affinit...e, \n    worker_logdir_root=None, \n    worker_port_range=(54000, 55000), \n    worker_ports=None, \n    working_dir=None\n).connected_managers
parsl/tests/test_scaling/test_scale_down.py:66: AssertionError
```

**To Reproduce**
unlikely - this looks like a low probability race condition to me.

**Expected behavior**
test should pass

**Environment**
CI, PR #1884

actually, a variation, in test_scale_out, not test_scale_down:

```
=================================== FAILURES ===================================
________________________________ test_scale_out ________________________________

    @pytest.mark.local
    def test_scale_out():
        dfk = parsl.dfk()
    
        # Since we have init_blocks = 0, at this point we should have 0 managers
        print("Executor : ", dfk.executors['htex_local'])
        print("Before")
        print("Managers   : ", dfk.executors['htex_local'].connected_managers)
        print("Outstanding: \n", dfk.executors['htex_local'].outstanding)
        assert len(dfk.executors['htex_local'].connected_managers) == 0, "Expected 0 managers at start"
    
        fus = [sleeper(i) for i in [3, 3, 25, 25, 50]]
    
        for i in range(2):
            fus[i].result()
    
        # At this point, since we have 1 task already processed we should have atleast 1 manager
        print("Between")
        print("Managers   : ", dfk.executors['htex_local'].connected_managers)
        print("Outstanding: \n", dfk.executors['htex_local'].outstanding)
>       assert len(dfk.executors['htex_local'].connected_managers) == 5, "Expected 5 managers once tasks are running"
E       AssertionError: Expected 5 managers once tasks are running
E       assert 4 == 5
```

--- a/parsl/tests/conftest.py
+++ b/parsl/tests/conftest.py
@@ -1,12 +1,16 @@
 import importlib.util
+import itertools
 import logging
 import os
 import pathlib
+import time
+import types
 import signal
 import sys
 import tempfile
 import threading
 import traceback
+import typing as t
 from datetime import datetime
 from glob import glob
 from itertools import chain
@@ -290,3 +294,67 @@ def pytest_ignore_collect(path):
         return True
     else:
         return False
+
+
+def create_traceback(start: int = 0) -> t.Optional[types.TracebackType]:
+    """
+    Dynamically create a traceback.
+
+    Builds a traceback from the top of the stack (the currently executing frame) on
+    down to the root frame.  Optionally, use start to build from an earlier stack
+    frame.
+
+    N.B. uses `sys._getframe`, which I only know to exist in CPython.
+    """
+    tb = None
+    for depth in itertools.count(start + 1, 1):
+        try:
+            frame = sys._getframe(depth)
+            tb = types.TracebackType(tb, frame, frame.f_lasti, frame.f_lineno)
+        except ValueError:
+            break
+    return tb
+
+
+@pytest.fixture
+def try_assert():
+    def _impl(
+        test_func: t.Callable[[], bool],
+        fail_msg: str = "",
+        timeout_ms: float = 5000,
+        attempts: int = 0,
+        check_period_ms: int = 20,
+    ):
+        tb = create_traceback(start=1)
+        timeout_s = abs(timeout_ms) / 1000.0
+        check_period_s = abs(check_period_ms) / 1000.0
+        if attempts > 0:
+            for _attempt_no in range(attempts):
+                if test_func():
+                    return
+                time.sleep(check_period_s)
+            else:
+                att_fail = (
+                    f"\n  (Still failing after attempt limit [{attempts}], testing"
+                    f" every {check_period_ms}ms)"
+                )
+                exc = AssertionError(f"{str(fail_msg)}{att_fail}".strip())
+                raise exc.with_traceback(tb)
+
+        elif timeout_s > 0:
+            end = time.monotonic() + timeout_s
+            while time.monotonic() < end:
+                if test_func():
+                    return
+                time.sleep(check_period_s)
+            att_fail = (
+                f"\n  (Still failing after timeout [{timeout_ms}ms], with attempts "
+                f"every {check_period_ms}ms)"
+            )
+            exc = AssertionError(f"{str(fail_msg)}{att_fail}".strip())
+            raise exc.with_traceback(tb)
+
+        else:
+            raise AssertionError("Bad assert call: no attempts or timeout period")
+
+    yield _impl

--- a/parsl/tests/test_scaling/test_scale_down.py
+++ b/parsl/tests/test_scaling/test_scale_down.py
@@ -1,80 +1,97 @@
 import logging
-import parsl
-import pytest
 import time
-from parsl import python_app
 
+import pytest
+
+import parsl
+
+from parsl import File, python_app
 from parsl.providers import LocalProvider
 from parsl.channels import LocalChannel
-# from parsl.launchers import SimpleLauncher
 from parsl.launchers import SingleNodeLauncher
-
 from parsl.config import Config
 from parsl.executors import HighThroughputExecutor
 
 logger = logging.getLogger(__name__)
+
+_max_blocks = 5
+_min_blocks = 2
 
 
 def local_config():
     return Config(
         executors=[
             HighThroughputExecutor(
-                heartbeat_period=2,
-                heartbeat_threshold=6,
-                poll_period=1,
+                heartbeat_period=1,
+                heartbeat_threshold=2,
+                poll_period=100,
                 label="htex_local",
+                address="127.0.0.1",
                 max_workers=1,
                 provider=LocalProvider(
                     channel=LocalChannel(),
                     init_blocks=0,
-                    max_blocks=5,
-                    min_blocks=2,
+                    max_blocks=_max_blocks,
+                    min_blocks=_min_blocks,
                     launcher=SingleNodeLauncher(),
                 ),
             )
         ],
-        max_idletime=5,
+        max_idletime=0.5,
         strategy='htex_auto_scale',
     )
 
 
 @python_app
-def sleeper(t):
+def waiting_app(ident: int, inputs=()):
+    import pathlib
     import time
-    time.sleep(t)
+
+    # Approximate an Event by writing to files; the test logic will poll this file
+    with open(inputs[0], "a") as f:
+        f.write(f"Ready: {ident}\n")
+
+    # Similarly, use Event approximation (file check!) by polling.
+    may_finish_file = pathlib.Path(inputs[1])
+    while not may_finish_file.exists():
+        time.sleep(0.01)
 
 
 # see issue #1885 for details of failures of this test.
 # at the time of issue #1885 this test was failing frequently
 # in CI.
 @pytest.mark.local
-def test_scale_out():
-    logger.info("start")
+def test_scale_out(tmpd_cwd, try_assert):
     dfk = parsl.dfk()
 
-    logger.info("initial asserts")
-    assert len(dfk.executors['htex_local'].connected_managers()) == 0, "Expected 0 managers at start"
+    num_managers = len(dfk.executors['htex_local'].connected_managers())
+
+    assert num_managers == 0, "Expected 0 managers at start"
     assert dfk.executors['htex_local'].outstanding == 0, "Expected 0 tasks at start"
 
-    logger.info("launching tasks")
-    fus = [sleeper(i) for i in [15 for x in range(0, 10)]]
+    ntasks = 10
+    ready_path = tmpd_cwd / "workers_ready"
+    finish_path = tmpd_cwd / "workers_may_continue"
+    ready_path.touch()
+    inputs = [File(str(ready_path)), File(str(finish_path))]
 
-    logger.info("waiting for warm up")
-    time.sleep(15)
+    futs = [waiting_app(i, inputs=inputs) for i in range(ntasks)]
 
-    logger.info("asserting 5 managers")
-    assert len(dfk.executors['htex_local'].connected_managers()) == 5, "Expected 5 managers after some time"
+    while ready_path.read_text().count("\n") < _max_blocks:
+        time.sleep(0.5)
 
-    logger.info("waiting for all futures to complete")
-    [x.result() for x in fus]
+    assert len(dfk.executors['htex_local'].connected_managers()) == _max_blocks
 
-    logger.info("asserting 0 outstanding tasks after completion")
-    assert dfk.executors['htex_local'].outstanding == 0, "Expected 0 outstanding tasks after future completion"
+    finish_path.touch()  # Approximation of Event, via files
+    [x.result() for x in futs]
 
-    logger.info("waiting a while for scale down")
-    time.sleep(25)
+    assert dfk.executors['htex_local'].outstanding == 0
 
-    logger.info("asserting 2 managers remain")
-    assert len(dfk.executors['htex_local'].connected_managers()) == 2, "Expected 2 managers when no tasks, lower bound by min_blocks"
+    def assert_kernel():
+        return len(dfk.executors['htex_local'].connected_managers()) == _min_blocks
 
-    logger.info("test passed")
+    try_assert(
+        assert_kernel,
+        fail_msg=f"Expected {_min_blocks} managers when no tasks (min_blocks)",
+        timeout_ms=15000,
+    )

